odoo/tests/common.py

```diff
@@ -224,6 +225,64 @@ def assertQueryCount(self, default=0, margin=0, **counters):
+          - Comparison between many2one field id done by id. Empty comparison can be done using any falsy value.
+
+        :param records:               The records to compare.
+        :param expected_values:       List of dicts expected to be exactly matched in records
+        '''
+
+        def _compare_candidate(record, candidate):
+            ''' Return True if the candidate matches the given record '''
+            for field_name in candidate.keys():
+                record_value = record[field_name]
+                candidate_value = candidate[field_name]
+                field_type = record._fields[field_name].type
+                if field_type == 'monetary':
+                    # Compare monetary field.
+                    currency_field_name = record._fields[field_name]._related_currency_field
+                    record_currency = record[currency_field_name]
+                    if record_currency.compare_amounts(candidate_value, record_value)\
+                            if record_currency else candidate_value != record_value:
+                        return False
+                elif field_type in ('one2many', 'many2many'):
+                    # Compare x2many relational fields.
+                    # Empty comparison must be an empty list to be True.
+                    return record_value.ids == candidate_value
+                elif field_type == 'many2one':
+                    # Compare many2one relational fields.
+                    # Every falsy value is allowed to compare with an empty record.
+                    return record_value.id == candidate_value if (record_value or candidate_value) else True
+                elif (candidate_value or record_value) and record_value != candidate_value:
+                    # Compare others fields if not both interpreted as falsy values.
+                    return False
+            return True
+
+        def _format_message(records, expected_values):
+            ''' Return a formatted representation of records/expected_values. '''
+            all_records_values = records.read(expected_values[0].keys(), load=False)
+            msg1 = '\n'.join(pprint.pformat(dic) for dic in all_records_values)
+            msg2 = '\n'.join(pprint.pformat(dic) for dic in expected_values)
+            return 'Current values:\n\n%s\n\nExpected values:\n\n%s' % (msg1, msg2)
+
+        # if the length or both things to compare is different, we can already tell they're not equal
+        if len(records) != len(expected_values):
+            msg = 'Wrong number of records to compare: %d != %d.\n\n' % (len(records), len(expected_values))
+            self.fail(msg + _format_message(records, expected_values))
+
+        for index, record in enumerate(records):
+            if not _compare_candidate(record, expected_values[index]):
+                msg = 'Unexpected record found at index %d.\n\n' % index
```

Comment:
The error message is confusing to me.
The record is not "unexpected", it simply does not match the given values.
It is either not the expected record, or one of its fields has the wrong value.